pig-tests/test/org/partiql/pig/tests/SumConverterTests.kt

```diff
@@ -0,0 +1,97 @@
+package org.partiql.pig.tests
+
+import com.amazon.ionelement.api.ionBool
+import com.amazon.ionelement.api.ionInt
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.ArgumentsSource
+import org.partiql.pig.tests.generated.ToyLang
+import kotlin.test.assertEquals
+
+/**
+ * Tests sum converters and and demonstrates their use to convert ToyLang's AST into a text representation
+ * that is vaguely similar to OCaml.  This example converts from a PIG domain sum type to a string, but these
+ * conversions can be applied to just about anything.
+ *
+ * ToyLang's AST actually has two sum types: expr and operator.
+ *
+ * This example also demonstrates how both can be used together.
+ */
+class SumConverterTests {
+val toyOperatorConverter = object : ToyLang.Operator.Converter<String> {
+    override fun convertPlus(node: ToyLang.Operator.Plus): String = "+"
+    override fun convertMinus(node: ToyLang.Operator.Minus): String = "-"
+    override fun convertTimes(node: ToyLang.Operator.Times): String = "*"
+    override fun convertDivide(node: ToyLang.Operator.Divide): String = "/"
+    override fun convertModulo(node: ToyLang.Operator.Modulo): String = "%"
+}
+
+val toyExprConverter = object : ToyLang.Expr.Converter<String> {
+    override fun convertLit(node: ToyLang.Expr.Lit): String = "${node.value}"
+    override fun convertVariable(node: ToyLang.Expr.Variable): String = node.name.text
+    override fun convertNot(node: ToyLang.Expr.Not): String = "!${convert(node.expr)}"
+    override fun convertNary(node: ToyLang.Expr.Nary): String {
+        // This Converter<T> implementation isn't responsible for converting ToyLang.Operator instances.
+        // delegate to toyOperatorConverter for that.
+        val op = toyOperatorConverter.convert(node.op)
+        return node.operands.joinToString(" $op ") { convert(it) }
+    }
+
+    override fun convertLet(node: ToyLang.Expr.Let): String =
+        "let ${node.name.text} = ${convert(node.value)} in ${convert(node.body)}"
+
+    override fun convertFunction(node: ToyLang.Expr.Function): String =
+        "fun (${node.varName.text}) -> ${convert(node.body)}"
+}
+
+    class TestCase(val ast: ToyLang.Expr, val expectedStringRepresentation: String)
```

Comment:
Can we use a `data class` here?

Reply:
that does slightly improve their appearance in the test-runner. :heavy_check_mark: